### ISSUE TITLE
ttnn.tanh low PCC for specific inputs used in albert_v2_xlarge

### ISSUE THREAD

**Issue by amalbasaTT** [tenstorrent/tt-metal#17841]
### Ticket
[Link to Github Issue](https://github.com/tenstorrent/pytorch2.0_ttnn/issues/739)

**Related issue**
https://github.com/tenstorrent/tt-metal/issues/17721

**Describe the bug**
ttnn.tanh has low PCC for a specific input used in albert_v2_xlarge for masked lm. Happens at `ttnn_tanh = ttnn.tanh(ttnn_multiply_4, )`. PCC is 0.9942273950085615.
Input to ttnn.tanh has spec:
- input_values:  [axlarge_tanh_input.csv](https://github.com/user-attachments/files/18771167/axlarge_tanh_input.csv)

- shape: (1, 9, 8192)
- dtype: ttnn.bfloat16
- dlayout: ttnn.TILE_LAYOUT

**To Reproduce**
1. Install torch-ttnn as explained in https://github.com/tenstorrent/pytorch2.0_ttnn?tab=readme-ov-file 
2. Possition to pytorch2.0_ttnn directory
3. Generate `albert-xlarge-v2_code.py` and `albert-xlarge-v2_inputs.pickle` by running command:
```
pytest tests/models/albert/test_albert_masked_lm.py --gen_op_accuracy_tests
```
4. In `albert-xlarge-v2_code.py`, after `ttnn_tanh = ttnn.tanh(ttnn_multiply_4, )`, add `test_accuracy(tanh, ttnn_tanh)`. In function definition `test_accuracy`, change pcc to be 0.999.
5. Run `albert-xlarge-v2_code.py`:
```
python3 tests/autogen_accuracy_tests/albert-xlarge-v2_code.py
```
Low pcc error will be at `test_accuracy(tanh, ttnn_tanh)`

**Additional Content**
Functions to extract and load inputs to `ttnn_tanh`:
```
def save_tensor_to_csv(tensor, filename):
    import pandas as pd
    tensor = tensor.contiguous()
    if tensor.dtype == torch.bfloat16:
        tensor = tensor.to(torch.float32)
    if tensor.ndim == 0: 
        df = pd.DataFrame([tensor.item()])
    elif tensor.ndim == 1:
        df = pd.DataFrame(tensor.numpy())
    elif tensor.ndim == 2:
        df = pd.DataFrame(tensor.numpy())
    elif tensor.ndim == 3:
        df = pd.DataFrame(tensor.view(-1, tensor.size(-1)).numpy())
    elif tensor.ndim == 4: 
        df = pd.DataFrame(tensor.view(-1, tensor.size(-1)).numpy())
    else:
        raise ValueError("Unsupported tensor dimension")
    df.to_csv(filename, index=False)

def load_tensor_from_csv(filename, shape):
    import pandas as pd
    df = pd.read_csv(filename)
    data = df.values
    tensor = torch.tensor(data)
    tensor = tensor.view(shape)
    return tensor
```
Unit test:
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten

    
def load_tensor_from_csv(filename, shape):
    import pandas as pd
    df = pd.read_csv(filename)
    data = df.values
    tensor = torch.tensor(data)
    tensor = tensor.view(shape)
    return tensor


def run_tanh_tests(
    file_path,
    input_shape, 
    dtype, 
    dlayout,
    device,
):
    x = load_tensor_from_csv(file_path[0], input_shape[0]).to(torch.bfloat16)

    try:
        # get ref result
        ref_value = aten.tanh.default(x, )
        
        tt_x = ttnn.from_torch(x, dtype=dtype[0], layout=dlayout[0], device=device)

        tt_result = ttnn.tanh(tt_x)
        tt_result = ttnn.to_torch(tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    assert_with_pcc(ref_value, tt_result, 0.999)


test_sweep_args = [
    (
        ["path/to/tanh_input.csv"],
        [(1, 9, 8192)],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
    ),
]


@pytest.mark.parametrize(
    "file_path, input_shape, dtype, dlayout",
    (test_sweep_args),
)
def test_tanh(file_path, input_shape, dtype, dlayout, device):
    run_tanh_tests(file_path, input_shape, dtype, dlayout, device)

```

**Comment by ayerofieiev-tt** [comment]
@amalbasaTT , why is this PCC an issue? It is not 0.999, but are you sure that if we make it any higher that the model accuracy will be fixed?

**Comment by cmaryanTT** [comment]
This may be an issue at the edges of the range that we support, which may be ok. @umadevimcw please confirm where this occurs.

**Comment by amalbasaTT** [comment]
> [@amalbasaTT](https://github.com/amalbasaTT) , why is this PCC an issue? It is not 0.999, but are you sure that if we make it any higher that the model accuracy will be fixed?

Well the model is quite big, and I believe that any deviation from the PCC of 0.999 will grow in magnitude as it propagates trough the model.

**Comment by KalaivaniMCW** [comment]
Hi @amalbasaTT , 
Is there a way to test with different set of inputs to check if PCC improves ?
for the given inputs, the PCC goes upto 0.994 like you've mentioned (assert 0.9942273950085615 >= 0.999)

<img width="1105" alt="Image" src="https://github.com/user-attachments/assets/5f703335-a3fb-451e-a579-b6e0b241764f" />

Hi @cmaryanTT ,
    We have an open issue for `ttnn.tanh` https://github.com/tenstorrent/tt-metal/issues/15132 and by looking at the inputs in the csv, they all fall under this range (-3.5 to 3.5 ) I had mentioned in the issue. the tanh LLK uses a look up table

**Comment by amalbasaTT** [comment]
@KalaivaniMCW you mean with randomized inputs under range -3.5 to 3.5?

**Comment by amalbasaTT** [comment]
@ayerofieiev-tt I tested if fixing tanh would approve accuracy in albert-v2-base model (will test for others too). I replaced every `ttnn_tanh` in forward function like this:
```
  #ttnn_tanh = ttnn.tanh(ttnn_multiply_4, )
  ttnn_tanh_intermediate = aten.tanh.default(ttnn.to_torch(ttnn_multiply_4), )
  ttnn_tanh = ttnn.from_torch(ttnn_tanh_intermediate, device=device, layout = ttnn.TILE_LAYOUT, dtype = ttnn.bfloat16)
```
Accuracy went to 0.9435372886378103

**Comment by KalaivaniMCW** [comment]
> [@KalaivaniMCW](https://github.com/KalaivaniMCW) you mean with randomized inputs under range -3.5 to 3.5?

I meant lesser inputs in range (-3.5 to 3.5) and more inputs outside this range. - that way PCC should improve so that we can be sure that the issue is with the range (-3.5 to 3.5) which points to the look up table.

Even testing in different dtypes didn't help with the PCC so I suspect it is because of https://github.com/tenstorrent/tt-metal/issues/15132

**Comment by amalbasaTT** [comment]
@KalaivaniMCW  Unit test (csv file: [axlarge_tanh_input.csv](https://github.com/user-attachments/files/18801084/axlarge_tanh_input.csv)):
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten

    
def load_tensor_from_csv(filename, shape):
    import pandas as pd
    df = pd.read_csv(filename)
    data = df.values
    tensor = torch.tensor(data)
    tensor = tensor.view(shape)
    return tensor


def run_tanh_tests(
    file_path,
    input_shape, 
    dtype, 
    dlayout,
    device,
):
    x = load_tensor_from_csv(file_path[0], input_shape[0]).to(torch.bfloat16)
    x= torch.Tensor(size=input_shape[0]).uniform_(-100, 100)
    
    num_modified_el = 4096
    num_elements = x.numel() 
    random_indices = torch.randperm(num_elements)[:num_modified_el]
    lower_bound = -3.5
    upper_bound = 3.5
    random_values = torch.empty(num_modified_el // 2).uniform_(lower_bound - 10, lower_bound).tolist() + \
                    torch.empty(num_modified_el // 2).uniform_(upper_bound, upper_bound + 10).tolist()

    random_values = torch.tensor(random_values).to(torch.bfloat16)
    x = x.flatten()
    x[random_indices] = random_values
    x = x.reshape(input_shape[0])
    
    
    try:
        # get ref result
        ref_value = aten.tanh.default(x, )
        
        tt_x = ttnn.from_torch(x, dtype=dtype[0], layout=dlayout[0], device=device)

        tt_result = ttnn.tanh(tt_x)
        tt_result = ttnn.to_torch(tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    assert_with_pcc(ref_value, tt_result, 0.999)


test_sweep_args = [
    (
        ["tests/ttnn/python_api_testing/non_working_unit_tests/wormhole/axlarge_tanh_input.csv"],
        [(1, 9, 8192)],
        [ttnn.bfloat16],
        [ttnn.TILE_LAYOUT],
    ),
]


@pytest.mark.parametrize(
    "file_path, input_shape, dtype, dlayout",
    (test_sweep_args),
)
def test_tanh(file_path, input_shape, dtype, dlayout, device):
    run_tanh_tests(file_path, input_shape, dtype, dlayout, device)

```
PCC seems to imporve as i increase `num_modified_el`.
